libtransmission/torrent.h

```diff
@@ -711,39 +711,114 @@ struct tr_torrent final : public tr_completion::torrent_view
         return bandwidth_group_;
     }
 
+    [[nodiscard]] constexpr auto peer_limit() const noexcept
+    {
+        return max_connected_peers_;
+    }
+
+    // --- idleness
+
+    constexpr void set_idle_limit_mode(tr_idlelimit mode) noexcept
+    {
+        if (mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED)
+        {
+            if (idle_limit_mode_ != mode)
+            {
+                idle_limit_mode_ = mode;
+                set_dirty();
+            }
+        }
+    }
+
     [[nodiscard]] constexpr auto idle_limit_mode() const noexcept
     {
         return idle_limit_mode_;
     }
 
+    constexpr void set_idle_limit_minutes(uint16_t idle_minutes) noexcept
+    {
+        if ((idle_limit_minutes_ != idle_minutes) && (idle_minutes > 0))
+        {
+            idle_limit_minutes_ = idle_minutes;
+            set_dirty();
+        }
+    }
+
     [[nodiscard]] constexpr auto idle_limit_minutes() const noexcept
     {
         return idle_limit_minutes_;
     }
 
-    [[nodiscard]] constexpr auto peer_limit() const noexcept
+    [[nodiscard]] constexpr std::optional<size_t> idle_seconds_left(time_t now) const noexcept
     {
-        return max_connected_peers_;
+        auto const idle_limit_minutes = effective_idle_limit_minutes();
+        if (!idle_limit_minutes)
+        {
+            return {};
+        }
+
+        auto const idle_seconds = this->idle_seconds(now);
+        if (!idle_seconds)
+        {
+            return {};
+        }
+
+        auto const idle_limit_seconds = size_t{ *idle_limit_minutes } * 60U;
+        return idle_limit_seconds > *idle_seconds ? idle_limit_seconds - *idle_seconds : 0U;
     }
 
-    constexpr void set_ratio_mode(tr_ratiolimit mode) noexcept
+    // --- seed ratio
+
+    constexpr void set_seed_ratio_mode(tr_ratiolimit mode) noexcept
     {
-        if (ratioLimitMode != mode)
+        if (mode == TR_RATIOLIMIT_GLOBAL || mode == TR_RATIOLIMIT_SINGLE || mode == TR_RATIOLIMIT_UNLIMITED)
```

Comment:
ditto

```cpp
TR_ASSERT(mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED);
```

Reply:
restored in 4e07e7bfcca6e0896d2270ccc8e9f14edd1f4c14